### PR TITLE
Another robustness fix for the PTE solvers

### PR DESCRIPTION
**Pull request by Yurlungur**
While debugging riot, I discovered another failure mode for the PTE solvers. In particular, the code has routines to enforce that volume fractions are initialized such that no materials start in a state such that dP/drho < 0. This can happen if you are to the expansion side of the minimum pressure in tension. If you start in this state, a Newton solver can't get you out of it---minimum pressure is a local extremum. 

This routine `SetVfracFromT` *does* set volume fractions to be outside of the spinodal region/vapor dome (where this can happen), but then it re-normalizes them so they sum to 1 (or vfrac_sum if specified != 1), and this can put you back into a bad state.

The proper thing to do is to use a [water filling algorithm](https://en.wikipedia.org/wiki/Water-filling_algorithm) that properly distributes the volume of a mixture across all materials while respecting their bounds. The spinodal requirement can be interpreted as a maximum allowable volume fraction for a given material.

I implement this change, and indeed it improves robustness. Some states that previously failed now succeed. Note that in some sense this was a bug. The solver was correct, but our initial guess machinery was sometimes choosing a pathological initial guess!

I still need to add a unit test enforcing that this mixture passes PTE. An interesting feature of this case is it's a mixture between an ideal gas and a tabulated EOS.

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [ ] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [ ] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.
- [ ] Ensure that any `when='@main'` dependencies are updated to the release version in the package.py

### CODE REVIEW

File: singularity-eos/closure/mixed_cell_models.hpp [hunk 1]
```diff
@@ -142,8 +142,9 @@ bool solve_Ax_b_wscr(const std::size_t n, Real *a, Real *b, Real *scr) {
 #warning "Eigen should not be used with Kokkos."
 #endif
   // Eigen VERSION
-  Eigen::Map<Eigen::Matrix<Real, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>> A(a, n,
-                                                                                     n);
+  using Matrix_t = Eigen::Matrix<Real, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+  Eigen::Map<Matrix_t> A(a, n, n);
```
Comment (Yurlungur) [inline singularity-eos/closure/mixed_cell_models.hpp:146]: this is just a legibility thing. Pay no heed.

File: singularity-eos/closure/mixed_cell_models.hpp [hunk 2]
```diff
@@ -248,19 +249,89 @@ class PTESolverBase {
   }
 
   PORTABLE_FORCEINLINE_FUNCTION
-  void SetVfracFromT(const Real T) {
-    Real vsum = 0.0;
-    // set volume fractions
+  void NormalizeVfrac() {
```
Comment (Yurlungur) [inline singularity-eos/closure/mixed_cell_models.hpp:252]: We call this enough times I figure we should have a function for it.

File: singularity-eos/closure/mixed_cell_models.hpp [hunk 3]
```diff
@@ -248,19 +249,89 @@ class PTESolverBase {
...
+   *
+   * If these are not all possible to respect, error out and expect
+   * the GetTguess function to save us by picking a higher
+   * temperature.
+   *
+   * If this is possible to achieve, we can do so by using the "water
+   * filling" algorithm. See, e.g., Gallager, 1968.
+   */
+  PORTABLE_FORCEINLINE_FUNCTION
+  bool SetVfracFromT(const Real T) {
```
Comment (Yurlungur) [inline singularity-eos/closure/mixed_cell_models.hpp:276]: The diff is kind of confusing here. I think just try to ignore the red. I completely rewrote this function.

**Comment by Yurlungur** [comment]
> For the specific case of a metal and gas, I agree that this code should help avoid tension states for the initial guess, but I'm not sure I actually understand the issue with the tension states.
> 
> For a globally non-convex EOS like a sesame table, I think we can all agree that the non-physical concave states should be avoided (i.e. those under the spinodal curve) since they can possibly provide an unphysical PTE solution, and I think that's actually what we're trying to preserve here rather than eliminating tension states. I can see how the spinodal states would represent a local minimum, but I don't understand how tension states broadly would be a local minimum.
> 
> The simplest counter-example is a collection of stiff gas EOS. Their minimum pressures occur at zero density so this mechanism would in effect do nothing. I also don't see how they'd have multiple solutions since the EOS is monotonic in pressure (although the energy isn't, and maybe that can mess you up in certain circumstances).
> 
> I'd rephrase much of this to reflect the fact that we're avoiding potential spinodal states (and also metal vapor states as an unfortunate consequence). If we wanted to truly avoid tension states, then the maximum volume fraction would reflect the density at zero pressure instead.

You're right, I mis-spoke. We're not avoiding tension, but rather the point of minimum pressure. I'll rephrase the relevant comments.

File: singularity-eos/closure/mixed_cell_models.hpp [hunk 4]
```diff
@@ -248,19 +249,95 @@ class PTESolverBase {
...
+    constexpr bool FAIL = true;
+    constexpr bool SUCCESS = false;
+    constexpr Real min_dv = 1e-12;
+
+    // TODO(JMM): Stash this somewhere rather than recomputing it?
+    auto get_vmax = [=](const std::size_t m) {
```
[the remark this answers is not in the document]
Comment (Yurlungur) [reply singularity-eos/closure/mixed_cell_models.hpp:282]: yeah I'll change it everywhere in the function as you suggest below.

File: singularity-eos/closure/mixed_cell_models.hpp [hunk 5]
```diff
@@ -248,19 +249,95 @@ class PTESolverBase {
...
+      return dv;
+    };
+
+    // Do an initial sweep of volume fractions to account for missing
+    // initial guesses
+    Real vmax_sum = 0;
     for (std::size_t m = 0; m < nmat; ++m) {
-      vfrac[m] *= robust::ratio(vfrac_total, vsum);
+      const Real vmax = get_vmax(m);
+      if (vfrac[m] <= 0) vfrac[m] = vmax;
```
[the remark this answers is not in the document]
Comment (Yurlungur) [reply singularity-eos/closure/mixed_cell_models.hpp:299]: This is basically duplicating old functionality. I *thought* that the purpose was to catch volume fractions where no valid initial guess was passed in. But now that I think about it, I'm not sure that's possible, since then rhobar wouldn't be well defined. 

Normal density isn't something we expose in the EOS API, unfortunately... but vmax will produce as low of a microphysical density as possible, right? Because specific volume is inversely proportional to density?

File: singularity-eos/closure/mixed_cell_models.hpp [hunk 6]
```diff
@@ -248,19 +249,95 @@ class PTESolverBase {
...
+      // Tally this to check if a valid non-tension state is possible
+      vmax_sum += vmax;
+    }
+    if (vmax_sum < vfrac_total) {
+      // no way to keep all materials out of tension with a volume
+      // fraction that sums to 1
+      return FAIL;
+    }
```
[the remark this answers is not in the document]
Comment (Yurlungur) [reply singularity-eos/closure/mixed_cell_models.hpp:307]: You're right, will rephrase comment.

File: singularity-eos/closure/mixed_cell_models.hpp [hunk 7]
```diff
@@ -311,15 +388,21 @@ class PTESolverBase {
       for (std::size_t m = 0; m < nmat; ++m) {
         Tguess = std::max(eos[m].MinimumTemperature(), Tguess);
       }
-      SetVfracFromT(Tguess);
+      rho_fail = SetVfracFromT(Tguess);
```
[the remark this answers is not in the document]
Comment (Yurlungur) [reply singularity-eos/closure/mixed_cell_models.hpp:391]: Switched to `bad_vfrac_guess`.

**Comment by pdmullen** [comment]
> > LGTM. To channel @jhp-lanl, does _this_ changeset fix the commented out PTE test? 😅
> 
> Yep it does. I uncommented it.

Neat.

File: singularity-eos/closure/mixed_cell_models.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhp-lanl) [reply singularity-eos/closure/mixed_cell_models.hpp]: > Normal density isn't something we expose in the EOS API, unfortunately...

Isn't that basically what `ValuesAtReferenceState()` does? It is exposed in the [API](https://github.com/lanl/singularity-eos/blob/main/singularity-eos/eos/eos_variant.hpp#L299).

Curiously, we also have the `ReferenceDensityTemperature()` function that isn't defined for all our EOS...

[the remark this answers is not in the document]
Comment (jhp-lanl) [reply singularity-eos/closure/mixed_cell_models.hpp]: > I'll make a comment that I'm not sure this can even happen, but let's leave it for now. As I think it preserves functionality originally here.

Sounds good. I think it's an edge case (or impossible case) anyway...

File: test/test_closure_pte.cpp [hunk 8]
```diff
@@ -160,6 +160,86 @@ bool run_PTE_from_state(const int num_pte, EOS *v_EOS, const Real spvol_bulk,
   return pte_converged;
 }
 
+constexpr int NT = 1;
+inline int RunPTE2Mat(EOS eos1, EOS eos2, Real rhobar1, Real rhobar2, Real alpha_guess1,
```
[the remark this answers is not in the document]
Comment (jhp-lanl) [reply test/test_closure_pte.cpp]: No worries... it's not that important and I can see why my helper function isn't as useful for a code like RIOT.

**Review by jhp-lanl — APPROVED** [review]

**Comment by jhp-lanl** [comment]
With the fix to the other failing PTE test, we may want to put out a release 🥳

**Comment by Yurlungur** [comment]
All tests pass merging!

Note, I found one other issue which I am fixing in another branch, so there will be one more follow up MR in this direction. Then maybe yeah, time for a release.
